### DISCUSSION TITLE
Should xcuitest delete the original video file on the phone after screen recording ends? Otherwise, will it result in disk garbage file overflow?

### DISCUSSION THREAD

**Discussion by youngfreeFJS** [appium/appium#20827] [Q&A] [answered]
Currently, only the files pulled down have been deleted, and I do not see the files displayed by the command xcrun devicectl device info files being deleted. @mykola-mokhnach @KazuCocoa 

![image](https://github.com/user-attachments/assets/1d4bfb35-bd8e-45a3-a9fc-d790e90b3a61)
![image](https://github.com/user-attachments/assets/267e042d-f7f9-4a38-a38f-725cc2df033c)

**Comment by KazuCocoa** [comment]
As, https://appium.github.io/appium-xcuitest-driver/latest/reference/execute-methods/#mobile-stopxctestscreenrecording currently we need to stop the XCTest session to delete files managed by the XCTest session for real devices. It means the WDA process must be ended https://github.com/appium/WebDriverAgent/blob/deb8ecabc52571e8004bcb2b804b9f7e43fad944/WebDriverAgentLib/Routing/FBWebServer.m#L233-L236

This is Apple's design. This is our findings in the past https://github.com/appium/WebDriverAgent/pull/858, so if the `xcrun devicectl device` starts providing a delete file command, it will be able to use it instead to delete files without XCTest session ends. I haven't checked the command recently, but do you find the way to delete files via `xcrun devicectl device`?

**Reply by youngfreeFJS** [reply]
@KazuCocoa Unfortunately, I also couldn't find the delete method for xcrun. 
Additionally, after executing the /wda/shutdown API, the video files on the phone still exist.

step1. run command `xcrun devicectl device info files --domain-type appDataContainer --domain-identifier com.apple.testmanagerd --username mobile --subdirectory Attachments --device 00008101-000C30261122001E`

return 
```
(venv) ➜  wda4ios17 xcrun devicectl device info files --domain-type appDataContainer --domain-identifier com.apple.testmanagerd --username mobile --subdirectory Attachments --device 00008101-000C30261122001E
14:49:44  Acquired tunnel connection to device.
14:49:44  Enabling developer disk image services.
14:49:44  Acquired usage assertion.
2 files:
Name
------------------------------------
542410C8-3D22-4A5D-B5F0-2B5271606E46
BA816847-5B62-44F3-AFC5-B3FDE0F43949
```

step2. curl shutdown API. `curl http://localhost:8100/wda/shutdown` (via WDA version 8.1.0)

step3. check file is delete or not
```
xcrun devicectl device info files --domain-type appDataContainer --domain-identifier com.apple.testmanagerd --username mobile --subdirectory Attachments --device 00008101-000C30261122001E

return:
(venv) ➜  wda4ios17 xcrun devicectl device info files --domain-type appDataContainer --domain-identifier com.apple.testmanagerd --username mobile --subdirectory Attachments --device 00008101-000C30261122001E
14:51:17  Enabling developer disk image services.
14:51:17  Acquired usage assertion.
2 files:
Name
------------------------------------
542410C8-3D22-4A5D-B5F0-2B5271606E46
BA816847-5B62-44F3-AFC5-B3FDE0F43949
```

**Reply by KazuCocoa** [reply]
Thanks. Then, it could also have some additional conditions internally I guess. As we also addressed _device factory reset_, we may eventually need to do a factory reset (erace all contents) for now. Hopefully, Apple will add a content deletion method via `devicectl`.

**Reply by youngfreeFJS** [reply]
> Thanks. Then, it could also have some additional conditions internally I guess. As we also addressed _device factory reset_, we may eventually need to do a factory reset (erace all contents) for now. Hopefully, Apple will add a content deletion method via `devicectl`.

`factory reset (erace all contents) for now` means real device factory reset right? By the way, is there any Apple documents or websites for latest xcrun commands? @KazuCocoa

**Reply by KazuCocoa** [reply, marked as answer]
Yes, it is settings -> general -> transfer or reset iPhone -> erace all content and settings in the latest version. It could vary iOS versions.
I don't know. We check CLI directly only. I think they are command help basis, so no such a official website/page

**Comment by youngfreeFJS** [comment]
Confirmed that there is still no method to delete files in the latest Xcode version (Xcode 16.1).

```
SUBCOMMANDS:
  copy                    Copy files.
  info                    Commands that provide information about a specific device
  install                 Install content onto a device.
  notification            Post and observe Darwin notifications on a device.
  orientation             Control device orientation.
  process                 Interact with processes on devices.
  reboot                  Reboot a device.
  uninstall               Uninstall content from a device.

  See 'device help <subcommand>' for detailed help.
➜  ~ xcrun devicectl device --version
397.24
➜  ~ xcode-select --version
xcode-select version 2409.
```

**Comment by youngfreeFJS** [comment]
The method you previously mentioned about deleting files by copying an empty file worked for me :) @KazuCocoa 

Example:
![image](https://github.com/user-attachments/assets/eaab033d-209a-4362-90af-e45d63ee49e6)

**Reply by mykola-mokhnach** [reply]
I assume it is possible to copy a new file, but devicectl does not allow to replace an existing one

**Reply by youngfreeFJS** [reply]
You are right!  @mykola-mokhnach cc @KazuCocoa 

I test record video 50 times, and then replace mobile video file as an empty file from computer,  via `xcrun devicectl device copy to` command.

Next, I pulled the "empty video" from my phone to my PC using the `xcrun DeviceCTL device info files`(check list) with `xcrun DeviceCTL device copy from `(pull file) commands, and the size of the video pulled from my computer was 0 (ls -al)
```
ufo@U-R4TNW2J1-1937 testvideopull % xcrun devicectl device info files --domain-type appDataContainer --domain-identifier com.apple.testmanagerd --username mobile --subdirectory Attachments --device 00008020-000E31982250003A
14:52:41  Acquired tunnel connection to device.
14:52:41  Enabling developer disk image services.
14:52:42  Acquired usage assertion.
49 files:
Name
------------------------------------
------------------------------------
00D9E891-D169-471D-B1C8-483285BED9D0
03420761-CF70-45ED-A165-8586DC41B69C
0C1A24E0-B822-40B1-B3BD-CCC71389C941
......

------------

ufo@U-R4TNW2J1-1937 testvideopull % xcrun devicectl device copy from --domain-type appDataContainer --domain-identifier com.apple.testmanagerd --source Attachments/00D9E891-D169-471D-B1C8-483285BED9D0 --destination ~/.pyufo/testvideopull/a.mov --user mobile --device 00008020-000E31982250003A
14:51:36  Acquired tunnel connection to device.
14:51:36  Enabling developer disk image services.
14:51:36  Acquired usage assertion.

----------------

ufo@U-R4TNW2J1-1937 testvideopull % ls -al
total 0
drwxr-xr-x  3 ufo  staff   96 12 10 14:43 .
drwxr-xr-x  9 ufo  staff  288 12 10 14:43 ..
-rw-r--r--@ 1 ufo  staff    0 12 10 14:43 a.mov
ufo@U-R4TNW2J1-1937 testvideopull %
```

But, in fact, when I checked the disk usage on my phone(Settings -> General -> iPhone Storage), it increased from 36GB/64GB to 38GB/64GB, and the usage of any apps in the details did not increase ...

**May I have any other way to further investigate or delete these files (excluding factory reset ...)?**

**Reply by KazuCocoa** [reply]
No, unfortunately. The only way is to wish Apple would add such a subcommand or we probably could find a way to submit a feature request to Apple

**Reply by youngfreeFJS** [reply]
> No, unfortunately. The only way is to wish Apple would add such a subcommand or we probably could find a way to submit a feature request to Apple

I have submitted a ticket to Apple and hope to receive their response ：）

**Reply by mykola-mokhnach** [reply]
I believe there is a way to delete such files same way xcode does that, however this API is private and the only way to get it working is via reverse engineering.

Maybe maintainers of https://github.com/doronz88/pymobiledevice3 or https://github.com/danielpaulus/go-ios would have more ideas

**Reply by KazuCocoa** [reply]
ah. The code below could also help; it needs to be modified to try it out, though. I think... `com.apple.afc` service still works.
https://github.com/appium/appium-xcuitest-driver/blob/7b293763a5ca3f4a526b76a0184f452b718bc2b4/lib/ios-fs-helpers.js#L263
https://github.com/appium/appium-ios-device/blob/09b2342ed839c0008d07ed19e786e23a9fbc6938/lib/afc/index.js#L71C9-L71C24

**Reply by mykola-mokhnach** [reply]
> The code below could also help; it needs to be modified to try it out, though. I think... com.apple.afc service still works.

When I tried that the testmanagerd container was not available for R/W via AFC service.

**Reply by youngfreeFJS** [reply]
I’m a bit confused. The video file I pulled was actually 0 bytes, but why is there a big increase in my phone’s storage space?

I'm sure that every mobile file in devicectl list, I pulled was 0 bytes.
